packages/astro/src/runtime.ts

```diff
@@ -137,6 +155,10 @@ async function load(config: RuntimeConfig, rawPathname: string | undefined): Pro
       let data: any[] = await loadData({ params: currentParams });
       if (!data) throw new Error(`[createCollection] \`data()\` returned nothing (empty data)"`);
       if (!Array.isArray(data)) data = [data]; // note: this is supposed to be a little friendlier to the user, but should we error out instead?
+      data = data.filter(entry => !entry.hasOwnProperty('published') || (entry.hasOwnProperty('published') && entry.published));
```

Comment:
I think that we'd actually want to do this inside of `Astro.fetchContent()`. With the new collections API, collections `data` isn't guarenteed to be Markdown content.